### PR TITLE
Optimize initializer modifier

### PR DESCRIPTION
**Pull request by k06a**
We have a hierarchy with more than 10 contracts, this optimization allowed us to significantly reduce gas cost

### CODE REVIEW

**Comment by spalladino** [comment]
@frangio can I ask you for your input here? The change seems harmless, and should help save gas, but I want to make sure I'm not missing anything.

File: packages/lib/contracts/Initializable.sol [hunk 1]
```diff
@@ -32,12 +32,18 @@ contract Initializable {
     require(initializing || isConstructor() || !initialized, "Contract instance has already been initialized");
 
     bool wasInitializing = initializing;
-    initializing = true;
-    initialized = true;
+    if (!wasInitializing) {
```
Comment (frangio) [inline packages/lib/contracts/Initializable.sol]:
```suggestion
    if (!initializing) {
```

This is more straightforward, less confusing, and makes it clear that this `if` block and the next one are doing exactly the same thing.

Comment (frangio) [inline packages/lib/contracts/Initializable.sol]: I personally think that `if (initializing != true)` would result in more self-evident code. What do you guys think?

**Comment by k06a** [comment]
Maybe you would also like this version:
```solidity
if (!initializing) {
  initialized = true;
  initializing = true;
  _;
  initializing = false;
} else {
  _;
}
```

**Comment by k06a** [comment]
image
https://blog.polymath.network/solidity-tips-and-tricks-to-save-gas-and-reduce-bytecode-size-c44580b218e6

**Comment by k06a** [comment]
@frangio what do you think?

**Comment by frangio** [comment]
It seems like `_` will duplicate the code. I'd keep the PR's original proposal.

**Comment by k06a** [comment]
@frangio fixed! Any suggestions?

**Comment by spalladino** [comment]
> Note that these optimizations won't be as significant after Constantinople, due to [EIP 1283](https://eips.ethereum.org/EIPS/eip-1283). However, I still think this is worth merging.

@frangio given 1283 was reverted, do you have any new considerations on this PR?

**Comment by frangio** [comment]
It makes it even more worth merging then.

**Comment by k06a** [comment]
Ok, I'm ready if you're ready :)

**Comment by spalladino** [comment]
Merged! Thanks a lot @k06a for the contribution!

**Pull request merged**